src/components/periodic-report/migrations/AddFinalReport.migration.ts

```diff
@@ -1,8 +1,74 @@
+import { node, relation } from 'cypher-query-builder';
 import { BaseMigration, Migration } from '../../../core';
+import { ReportType } from '../dto';
 
 @Migration('2021-08-03T15:30:05')
 export class AddFinalReportMigration extends BaseMigration {
   async up() {
-    // TODO implement
+    await this.migrate(ReportType.Financial);
+    await this.migrate(ReportType.Narrative);
+    await this.migrate(ReportType.Progress);
+  }
+
+  // the finalReportMigration property is present on all the migrated :PeriodicReport nodes if there is a need to revert
+  private async migrate(type: ReportType) {
+    const before = await this.getBaseNodeCount(type, false);
+    this.logger.info(
+      `${before ?? 0} ${
+        type !== ReportType.Progress ? 'projects' : 'engagements'
+      } with ${type} reports before migration`
+    );
+    await this.db
+      .query()
+      .raw(
+        `
+          match(u:User)-[:email { active: true }]->(:Property { value: "devops@tsco.org" })
+          match(b:BaseNode)-[:report]->(:${type}Report)
+          with u, collect(distinct b) as baseNodes
+          unwind baseNodes as baseNode
+          call {
+            with baseNode
+            match(baseNode)-[:report]->(:${type}Report)-[:end]->(en:Property)
+            with en
+            order by en.value desc limit 1
+            return en.value as lastEnd
+          }
```

Comment:
I know I said previously that we don't need to filter on reports that already have a final report. But now I'm thinking it might be better safe than sorry...